lib/connection/pool.js

```diff
@@ -1007,25 +1049,49 @@ Pool.prototype.write = function(commands, options, cb) {
     operation.socketTimeout = options.socketTimeout;
   }
 
-  // We need to have a callback function unless the message returns no response
-  if(!(typeof cb == 'function') && !options.noResponse) {
-    throw new MongoError('write method must provide a callback');
+  // Ensure commands is an array
+  if (!Array.isArray(commands)) {
+    commands = [commands];
   }
 
-  // If we have a monitoring operation schedule as the very first operation
-  // Otherwise add to back of queue
-  if(options.monitoring) {
-    this.queue.unshift(operation);
-  } else {
-    this.queue.push(operation);
-  }
+  // Get the requestId
+  operation.requestId = commands[commands.length - 1].requestId;
 
-  // Attempt to execute the operation
-  if(!self.executing) {
-    process.nextTick(function() {
-      _execute(self)();
-    });
-  }
+  // Prepare the operation buffer
+  serializeCommands(self, commands, [], function(err, serializedCommands) {
+    if (err) throw err;
+
+    // Set the operation's buffer to the serialization of the commands
+    operation.buffer = serializedCommands;
+
+    // If we have a monitoring operation schedule as the very first operation
+    // Otherwise add to back of queue
+    if(options.monitoring) {
+      self.queue.unshift(operation);
+    } else {
+      self.queue.push(operation);
+    }
+
+    // Attempt to execute the operation
+    if(!self.executing) {
+      process.nextTick(function() {
+        _execute(self)();
+      });
+    }
+  })
+
+}
+
+// Return whether a command contains an uncompressible command term
+// Will return true if command contains no uncompressible command terms
+var checkForUncompressibleCommands = function(command) {
```

Comment:
@malexandert I think you wrote this function originally – what do you want to do about this suggestion?